### PR TITLE
texlive: 2021.20210408 -> 2021.20211227

### PR DESCRIPTION
**Pull request by veprbl**
###### Motivation for this change

Preparation for shipping TeXLive 2021-final in NixOS 22.05

###### Things done

I was using this for some time locally.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Closes: #126714

### CODE REVIEW

**Comment by veprbl** [comment]
@GrahamcOfBorg build tests.texlive

**Comment by lovesegfault** [comment]
Result of `nixpkgs-review pr 154105` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>gnome.gnome-documents</li>
    <li>hyper-haskell</li>
    <li>octave-jit</li>
    <li>octavePackages.fem-fenics</li>
    <li>octavePackages.vibes</li>
    <li>octavePackages.vrml</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.cnvkit</li>
    <li>python310Packages.fipy</li>
    <li>python310Packages.pplpy</li>
    <li>python39Packages.cnvkit</li>
  </ul>
</details>
<details>
  <summary>259 packages built:</summary>
  <ul>
    <li>R</li>
    <li>adapta-gtk-theme</li>
    <li>afterstep</li>
    <li>alsa-tools</li>
    <li>apostrophe</li>
    <li>apt</li>
    <li>asciidoc-full</li>
    <li>asciidoc-full-with-plugins</li>
    <li>asymptote</li>
    <li>auctex</li>
    <li>auto-multiple-choice</li>
    <li>autofs5</li>
    <li>biber</li>
    <li>bibletime</li>
    <li>blahtexml</li>
    <li>bluespec</li>
    <li>btrbk</li>
    <li>bumblebee</li>
    <li>cddlib</li>
    <li>chrome-gnome-shell</li>
    <li>clevis</li>
    <li>crowbar</li>
    <li>csound</li>
    <li>csound-qt</li>
    <li>cubicsdr</li>
    <li>dblatex</li>
    <li>dblatexFull</li>
    <li>denemo</li>
    <li>diffoscope</li>
    <li>digikam</li>
    <li>dillo</li>
    <li>disorderfs</li>
    <li>dot2tex (python39Packages.dot2tex)</li>
    <li>dump1090</li>
    <li>dwarf-fortress-packages.dwarf-fortress-full</li>
    <li>dwarf-therapist (dwarf-fortress-packages.dwarf-therapist)</li>
    <li>dwarf-fortress-packages.dwarf-therapist-original</li>
    <li>eclib</li>
    <li>enblend-enfuse</li>
    <li>ensemble-chorus</li>
    <li>eureka-editor</li>
    <li>evince (gnome.evince)</li>
    <li>exrdisplay</li>
    <li>faust2csound</li>
    <li>fldigi</li>
    <li>fllog</li>
    <li>flmsg</li>
    <li>flpsed</li>
    <li>flrig</li>
    <li>fltk (fltk13)</li>
    <li>fltk14</li>
    <li>flwrap</li>
    <li>frescobaldi</li>
    <li>gama</li>
    <li>gfan</li>
    <li>giac</li>
    <li>giac-with-xcas</li>
    <li>giada</li>
    <li>glom</li>
    <li>gmsh</li>
    <li>gnome.gnome-books</li>
    <li>gnome.gnome-session (gnome.gnome_session)</li>
    <li>gnome.gnome-shell (gnome.gnome_shell)</li>
    <li>gnome.gnome-terminal (gnome.gnome_terminal)</li>
    <li>gnome.gnome-tweaks (gnome.gnome-tweak-tool)</li>
    <li>gnome.sushi</li>
    <li>gnomeExtensions.easyScreenCast</li>
    <li>gnomeExtensions.gsconnect</li>
    <li>gnomeExtensions.night-theme-switcher</li>
    <li>gnss-sdr</li>
    <li>gnuradio3_8Packages.ais</li>
    <li>gnuradio3_8Packages.limesdr</li>
    <li>gnuradio3_8Packages.osmosdr</li>
    <li>gp2c</li>
    <li>gqrx</li>
    <li>gqrx-gr-audio</li>
    <li>gqrx-portaudio</li>
    <li>gummi</li>
    <li>herwig</li>
    <li>hugin</li>
    <li>ipe</li>
    <li>ja2-stracciatella</li>
    <li>jwm-settings-manager</li>
    <li>rep (kakounePlugins.rep)</li>
    <li>lcalc</li>
    <li>librsb</li>
    <li>lilypond</li>
    <li>lilypond-with-fonts</li>
    <li>limesuite</li>
    <li>liquidwar</li>
    <li>lmms</li>
    <li>luksmeta</li>
    <li>manim</li>
    <li>mate.atril</li>
    <li>minc_widgets</li>
    <li>mitscheme</li>
    <li>mitschemeX11</li>
    <li>modem-manager-gui</li>
    <li>mup</li>
    <li>ne</li>
    <li>new-session-manager</li>
    <li>nuweb</li>
    <li>octave</li>
    <li>octaveFull</li>
    <li>octavePackages.arduino</li>
    <li>octavePackages.audio</li>
    <li>octavePackages.bim</li>
    <li>octavePackages.bsltl</li>
    <li>octavePackages.cgi</li>
    <li>octavePackages.communications</li>
    <li>octavePackages.control</li>
    <li>octavePackages.data-smoothing</li>
    <li>octavePackages.database</li>
    <li>octavePackages.dataframe</li>
    <li>octavePackages.dicom</li>
    <li>octavePackages.divand</li>
    <li>octavePackages.doctest</li>
    <li>octavePackages.econometrics</li>
    <li>octavePackages.financial</li>
    <li>octavePackages.fits</li>
    <li>octavePackages.fpl</li>
    <li>octavePackages.fuzzy-logic-toolkit</li>
    <li>octavePackages.ga</li>
    <li>octavePackages.general</li>
    <li>octavePackages.generate_html</li>
    <li>octavePackages.geometry</li>
    <li>octavePackages.gsl</li>
    <li>octavePackages.image</li>
    <li>octavePackages.image-acquisition</li>
    <li>octavePackages.instrument-control</li>
    <li>octavePackages.interval</li>
    <li>octavePackages.io</li>
    <li>octavePackages.level-set</li>
    <li>octavePackages.linear-algebra</li>
    <li>octavePackages.lssa</li>
    <li>octavePackages.ltfat</li>
    <li>octavePackages.mapping</li>
    <li>octavePackages.matgeom</li>
    <li>octavePackages.miscellaneous</li>
    <li>octavePackages.msh</li>
    <li>octavePackages.mvn</li>
    <li>octavePackages.nan</li>
    <li>octavePackages.ncarray</li>
    <li>octavePackages.netcdf</li>
    <li>octavePackages.nurbs</li>
    <li>octavePackages.ocl</li>
    <li>octavePackages.octclip</li>
    <li>octavePackages.octproj</li>
    <li>octavePackages.optics</li>
    <li>octavePackages.optim</li>
    <li>octavePackages.optiminterp</li>
    <li>octavePackages.parallel</li>
    <li>octavePackages.quaternion</li>
    <li>octavePackages.queueing</li>
    <li>octavePackages.signal</li>
    <li>octavePackages.sockets</li>
    <li>octavePackages.sparsersb</li>
    <li>octavePackages.splines</li>
    <li>octavePackages.statistics</li>
    <li>octavePackages.stk</li>
    <li>octavePackages.strings</li>
    <li>octavePackages.struct</li>
    <li>octavePackages.symbolic</li>
    <li>octavePackages.tisean</li>
    <li>octavePackages.tsa</li>
    <li>octavePackages.video</li>
    <li>octavePackages.windows</li>
    <li>octavePackages.zeromq</li>
    <li>openems</li>
    <li>openmolcas</li>
    <li>openwebrx</li>
    <li>pantheon.elementary-session-settings</li>
    <li>pantheon.evince</li>
    <li>paperwork</li>
    <li>pari</li>
    <li>paulstretch</li>
    <li>perl532Packages.Po4a</li>
    <li>perl534Packages.Po4a</li>
    <li>phosh</li>
    <li>polymake</li>
    <li>posterazor</li>
    <li>postgresql11Packages.plr</li>
    <li>postgresql12Packages.plr</li>
    <li>postgresqlPackages.plr (postgresql13Packages.plr)</li>
    <li>postgresql14Packages.plr</li>
    <li>pothos</li>
    <li>pynac</li>
    <li>python310Packages.cypari2</li>
    <li>python310Packages.cysignals</li>
    <li>python310Packages.dot2tex</li>
    <li>python310Packages.fpylll</li>
    <li>python310Packages.pypandoc</li>
    <li>python310Packages.python-csxcad</li>
    <li>python310Packages.python-openems</li>
    <li>python310Packages.rivet</li>
    <li>python310Packages.rpy2</li>
    <li>python310Packages.soapysdr-with-plugins</li>
    <li>python310Packages.sphinxcontrib-bayesnet</li>
    <li>python310Packages.sphinxcontrib-tikz</li>
    <li>python39Packages.cypari2</li>
    <li>python39Packages.cysignals</li>
    <li>python39Packages.fipy</li>
    <li>python39Packages.fpylll</li>
    <li>python39Packages.pplpy</li>
    <li>python39Packages.pypandoc</li>
    <li>python39Packages.python-csxcad</li>
    <li>python39Packages.python-openems</li>
    <li>rivet (python39Packages.rivet)</li>
    <li>python39Packages.rpy2</li>
    <li>python39Packages.soapysdr-with-plugins</li>
    <li>python39Packages.sphinxcontrib-bayesnet</li>
    <li>python39Packages.sphinxcontrib-tikz</li>
    <li>qradiolink</li>
    <li>rakarrack</li>
    <li>rstudio</li>
    <li>rstudioWrapper</li>
    <li>rtl_433</li>
    <li>sage</li>
    <li>sageWithDoc</li>
    <li>sdrangel</li>
    <li>seaview</li>
    <li>sherpa</li>
    <li>singular</li>
    <li>skribilo</li>
    <li>soapysdr-with-plugins</li>
    <li>solfege</li>
    <li>spark (spark3)</li>
    <li>spark2</li>
    <li>spring</li>
    <li>springLobby</li>
    <li>sssd</li>
    <li>sympow</li>
    <li>tang</li>
    <li>tectonic</li>
    <li>texlive.combined.scheme-basic</li>
    <li>texlive.combined.scheme-context</li>
    <li>texlive.combined.scheme-full</li>
    <li>texlive.combined.scheme-gust</li>
    <li>texlive.combined.scheme-infraonly</li>
    <li>texlive.combined.scheme-medium</li>
    <li>texlive.combined.scheme-minimal</li>
    <li>texlive.combined.scheme-small</li>
    <li>texlive.combined.scheme-tetex</li>
    <li>texmacs</li>
    <li>thepeg</li>
    <li>tigervnc</li>
    <li>udiskie</li>
    <li>urh</li>
    <li>vimix-gtk-themes</li>
    <li>virtualgl</li>
    <li>virtualglLib</li>
    <li>welle-io</li>
    <li>whitesur-gtk-theme</li>
    <li>xautoclick</li>
    <li>yoshimi</li>
    <li>yosys-bluespec</li>
    <li>zathura</li>
    <li>zettlr</li>
    <li>zynaddsubfx-fltk</li>
  </ul>
</details>

**Comment by lovesegfault** [comment]
All failures happen on master too, so this seems AOK

File: pkgs/tools/typesetting/tex/texlive/combine.nix [hunk 1]
```diff
@@ -223,17 +223,17 @@ in (buildEnv {
...
-    #${bin.texlinks} "$out/bin" && wrapBin # do we need to regenerate format links?
+    #${bin.texlinks}/bin/texlinks.sh "$out/bin" && wrapBin # do we need to regenerate format links?
 
     # Disable unavailable map files
     echo y | perl `type -P updmap.pl` --sys --syncwithtrees --force
     # Regenerate the map files (this is optional)
     perl `type -P updmap.pl` --sys --force
 
-    perl `type -P mktexlsr.pl` ./share/texmf-* # to make sure
+    perl `type -P mktexlsr.pl` --sort ./share/texmf-* # to make sure
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/combine.nix:236]: Cc #102917 – this is a partial fix towards making the combined schemes reproducible.

**Review by xworld21 — CHANGES_REQUESTED** [review]
Thanks! This looks good to me. Unfortunately I can't test it properly right now, but I don't see why this should be held back, it is quite harmless.

(It may cause a regression in LaTeXML, though.)

File: pkgs/tools/typesetting/tex/texlive/bin.nix [hunk 2]
```diff
@@ -373,7 +375,7 @@ texlinks = stdenv.mkDerivation rec {
     # Patch texlinks.sh back to 2015 version;
     # otherwise some bin/ links break, e.g. xe(la)tex.
     patch --verbose -R scripts/texlive-extra/texlinks.sh < '${./texlinks.diff}'
-    install -Dm555 scripts/texlive-extra/texlinks.sh "$out"
+    install -Dm555 scripts/texlive-extra/texlinks.sh "$out"/bin/texlinks.sh
```
[the remark this answers is not in the document]
Comment (veprbl) [reply pkgs/tools/typesetting/tex/texlive/bin.nix]: The name could be anything here, but, yes, things going from texmf/scripts to bin lose the filename extension.

File: pkgs/tools/typesetting/tex/texlive/bin.nix [hunk 3]
```diff
@@ -136,6 +136,8 @@ core = stdenv.mkDerivation rec {
   + /* doc location identical with individual TeX pkgs */ ''
     mkdir -p "$doc/doc"
     mv "$out"/share/{man,info} "$doc"/doc
+  '' + /* remove manpages for utils that live in texlive.texlive-scripts to avoid a conflict in buildEnv */ ''
+    (cd "$doc"/doc/man/man1; rm {fmtutil-sys.1,fmtutil.1,mktexfmt.1,mktexmf.1,mktexpk.1,mktextfm.1,texhash.1,updmap-sys.1,updmap.1})
```
[the remark this answers is not in the document]
Comment (veprbl) [reply pkgs/tools/typesetting/tex/texlive/bin.nix:140]: It does fix a regression in `tests.texlive.texdoc`. However the #118757 would require something like
```
diff --git a/pkgs/test/texlive/default.nix b/pkgs/test/texlive/default.nix
index 217a862e1c5..1109a73fb5f 100644
--- a/pkgs/test/texlive/default.nix
+++ b/pkgs/test/texlive/default.nix
@@ -109,7 +109,7 @@
   texdoc = runCommand "texlive-test-texdoc" {
     nativeBuildInputs = [
       (with texlive; combine {
-        inherit scheme-infraonly luatex texdoc;
+        inherit scheme-full;
         pkgFilter = pkg: lib.elem pkg.tlType [ "run" "bin" "doc" ];
       })
     ];
```
to cover all cases, and that requires more fixes.

File: pkgs/tools/typesetting/tex/texlive/combine.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/tools/typesetting/tex/texlive/combine.nix:236]: Right, I think several people, including you, had figured out this part, but it was never merged

**Comment by veprbl** [comment]
@GrahamcOfBorg build tests.texlive

**Review by xworld21 — APPROVED** [review]

File: pkgs/tools/typesetting/tex/texlive/fixedHashes.nix [hunk 4]
```diff
@@ -1,9095 +1 @@
-{
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/fixedHashes.nix:1]: @veprbl I didn't notice `fixedHashes.nix` was empty... is this intentional? (I am generating the April one right now, by the way.)

Comment (veprbl) [reply pkgs/tools/typesetting/tex/texlive/fixedHashes.nix:1]: Must be a bug. A fix would be appreciated.

**Pull request merged**